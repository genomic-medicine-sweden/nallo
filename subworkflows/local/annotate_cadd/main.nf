//
// A subworkflow to annotate snvs
//

include { BCFTOOLS_ANNOTATE             } from '../../../modules/nf-core/bcftools/annotate/main'
include { BCFTOOLS_VIEW                 } from '../../../modules/nf-core/bcftools/view/main'
include { CADD                          } from '../../../modules/nf-core/cadd/main'
include { TABIX_TABIX as TABIX_ANNOTATE } from '../../../modules/nf-core/tabix/tabix/main'
include { TABIX_TABIX as TABIX_CADD     } from '../../../modules/nf-core/tabix/tabix/main'
include { TABIX_TABIX as TABIX_VIEW     } from '../../../modules/nf-core/tabix/tabix/main'

workflow ANNOTATE_CADD {

    take:
        ch_vcf            // channel: [mandatory] [ val(meta), path(vcfs) ]
        ch_index          // channel: [mandatory] [ val(meta), path(tbis) ]
        ch_header         // channel: [mandatory] [ path(txt) ]
        ch_cadd_resources // channel: [mandatory] [ path(dir) ]

    main:
        ch_versions       = Channel.empty()

        BCFTOOLS_VIEW(ch_vcf.join(ch_index), [], [], [])

        TABIX_VIEW(BCFTOOLS_VIEW.out.vcf)

        CADD(BCFTOOLS_VIEW.out.vcf, ch_cadd_resources)

        TABIX_CADD(CADD.out.tsv)

        ch_vcf
            .join(ch_index)
            .join(CADD.out.tsv)
            .join(TABIX_CADD.out.tbi)
            .combine(ch_header)
            .set { ch_annotate_in }

        BCFTOOLS_ANNOTATE(ch_annotate_in)

        TABIX_ANNOTATE (BCFTOOLS_ANNOTATE.out.vcf)

        ch_versions = ch_versions.mix(BCFTOOLS_VIEW.out.versions.first())
        ch_versions = ch_versions.mix(TABIX_VIEW.out.versions.first())
        ch_versions = ch_versions.mix(CADD.out.versions.first())
        ch_versions = ch_versions.mix(TABIX_CADD.out.versions.first())
        ch_versions = ch_versions.mix(BCFTOOLS_ANNOTATE.out.versions.first())
        ch_versions = ch_versions.mix(TABIX_ANNOTATE.out.versions.first())

    emit:
        vcf  = BCFTOOLS_ANNOTATE.out.vcf // channel: [ val(meta), path(vcf) ]
        tbi  = TABIX_ANNOTATE.out.tbi    // channel: [ val(meta), path(tbi) ]
        versions = ch_versions           // channel: [ path(versions.yml) ]
}
