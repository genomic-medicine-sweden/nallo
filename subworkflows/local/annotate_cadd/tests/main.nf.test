nextflow_workflow {

    name "Test Workflow ANNOTATE_CADD"
    script "../main.nf"
    config "./nextflow.config"
    workflow "ANNOTATE_CADD"

    setup {
        run("GUNZIP") {
            script "../../../../modules/nf-core/gunzip/main.nf"
            process {
                """
                input[0] = [
                    [ id:'test' ],
                    file(params.pipelines_testdata_base_path + 'reference/hg38.test.fa.gz', checkIfExists: true)
                ]
                """
            }

        }
        run("SAMTOOLS_FAIDX") {
            script "../../../../modules/nf-core/samtools/faidx/main.nf"
            process {
                """
                input[0] = GUNZIP.out.gunzip
                input[1] = [[],[]]
                input[2] = false
                """
            }
        }
        run("DEEPVARIANT_RUNDEEPVARIANT") {
            script "../../../../modules/nf-core/deepvariant/rundeepvariant/main.nf"
            process {
                """
                input[0] = [
                    [ id: 'test', family_id: 'test', sex: 1 ],
                    file(params.pipelines_testdata_base_path + 'testdata/HG002_PacBio_Revio.bam', checkIfExists: true),
                    file(params.pipelines_testdata_base_path + 'testdata/HG002_PacBio_Revio.bam.bai', checkIfExists: true),
                    file(params.pipelines_testdata_base_path + 'reference/test_data.bed', checkIfExists: true)
                ]
                input[1] = GUNZIP.out.gunzip
                input[2] = SAMTOOLS_FAIDX.out.fai
                input[3] = [[],[]]
                input[4] = [[],[]]
                """
            }
        }
    }

    test("1 sample - stub") {
        options "-stub"

        when {
            workflow {
                """
                input[0] = SAMTOOLS_FAIDX.out.fai
                input[1] = DEEPVARIANT_RUNDEEPVARIANT.out.vcf
                input[2] = DEEPVARIANT_RUNDEEPVARIANT.out.vcf_index
                input[3] = Channel.of([
                    [ id: 'test' ],
                    file("$projectDir/assets/cadd_to_vcf_header_-1.0-.txt", checkIfExists: true)
                ])
                input[4] = Channel.of([
                    [ id: 'test' ],
                    file("$projectDir/assets/", checkIfExists: true)
                ])
                input[5] = Channel.of([
                    [ id: 'test' ],
                    file("$projectDir/docs/", checkIfExists: true)
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }
    }
}
