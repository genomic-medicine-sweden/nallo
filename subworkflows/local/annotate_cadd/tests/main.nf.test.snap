{
    "1 sample - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "family_id": "test",
                            "sex": "1"
                        },
                        "rename_chrnames_ann.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "family_id": "test",
                            "sex": "1"
                        },
                        "rename_chrnames_ann.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    "versions.yml:md5,2fafab6b15acf7fcae148e0dea83d28a",
                    "versions.yml:md5,43689963f80fac1f15861b144f2c217b",
                    "versions.yml:md5,457a734ccb40d752d1ed0f804dfdefa5",
                    "versions.yml:md5,8bd330c3d52c6227c294d82b35e39bdc",
                    "versions.yml:md5,9c99528881382100093a6b3d6813c6bb",
                    "versions.yml:md5,b869f1605e2f022a8d364d081fdc348e",
                    "versions.yml:md5,e731cf5c84718f20c8455f0fd2e8b89d"
                ],
                "tbi": [
                    [
                        {
                            "id": "test",
                            "family_id": "test",
                            "sex": "1"
                        },
                        "rename_chrnames_ann.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "vcf": [
                    [
                        {
                            "id": "test",
                            "family_id": "test",
                            "sex": "1"
                        },
                        "rename_chrnames_ann.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,2fafab6b15acf7fcae148e0dea83d28a",
                    "versions.yml:md5,43689963f80fac1f15861b144f2c217b",
                    "versions.yml:md5,457a734ccb40d752d1ed0f804dfdefa5",
                    "versions.yml:md5,8bd330c3d52c6227c294d82b35e39bdc",
                    "versions.yml:md5,9c99528881382100093a6b3d6813c6bb",
                    "versions.yml:md5,b869f1605e2f022a8d364d081fdc348e",
                    "versions.yml:md5,e731cf5c84718f20c8455f0fd2e8b89d"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-10-22T13:17:28.752327377"
    }
}